.github/workflows/publish.yml

```diff
@@ -0,0 +1,87 @@
+name: Publish Docker image
+on:
+# Will only run when release is published. 
+  release:
+    types: 
+      - created
+  workflow_dispatch:
+
+jobs:
+  publish-images:
+    runs-on: ubuntu-20.04
+    steps:
+      # Set fetch-depth: 0 to fetch commit history and tags for use in version calculation
+      - name: Check out code
+        uses: actions/checkout@v2.3.4
+        with:
+          fetch-depth: 0
+
+      - name: create checksum file
+        run: find . -type f -name "*.gradle*" -o -name "gradle-wrapper*" | xargs shasum | sort > checksum.txt && cat checksum.txt
+
+      - name: Cache packages
+        id: cache-packages
+        uses: actions/cache@v2
+        with:
+          path: ~/.gradle
+          key: gradle-packages-${{ runner.os }}-${{ hashFiles('**/checksum.txt') }}
+
+      - name: download dependencies
```

Comment:
So given how GHA works with caching, I think we can remove the download dependencies jobs entirely - they're meant to fill the cache before saving it. Instead, we can let dependencies download as needed and have GH save them at the end. Because the jobs are potentially using different dependencies, they'd overwrite each other if they used the same cache key. So we may want to add the job name as one of the segments of the key, and add another restore key match like
```
*-{os}-{job}
*-{os}
```